gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/ClassUniquenessAnalyzer.java

```diff
@@ -0,0 +1,155 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.jar.JarEntry;
+import java.util.jar.JarInputStream;
+import java.util.stream.Collectors;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.slf4j.Logger;
+
+public final class ClassUniquenessAnalyzer {
+
+    private final Map<Set<ModuleVersionIdentifier>, Set<String>> jarsToClasses = new HashMap<>();
+    private final Map<String, Set<HashCode>> classToHashCodes = new HashMap<>();
+    private final Logger log;
+
+    public ClassUniquenessAnalyzer(Logger log) {
+        this.log = log;
+    }
+
+    public void analyzeConfiguration(Configuration configuration) {
+        Instant before = Instant.now();
+        Set<ResolvedArtifact> dependencies = configuration
+                .getResolvedConfiguration()
+                .getResolvedArtifacts();
+
+        // we use these temporary maps to accumulate information as we process each jar,
+        // so they may include singletons which we filter out later
+        Map<String, Set<ModuleVersionIdentifier>> classToJars = new HashMap<>();
+        Map<String, Set<HashCode>> tempClassToHashCodes = new HashMap<>();
+
+        dependencies.stream().forEach(resolvedArtifact -> {
+            File file = resolvedArtifact.getFile();
+            if (!file.exists()) {
+                log.info("Skipping non-existent jar {}: {}", resolvedArtifact, file);
+                return;
+            }
+
+            try (FileInputStream fileInputStream = new FileInputStream(file);
+                    JarInputStream jarInputStream = new JarInputStream(fileInputStream)) {
+                JarEntry entry;
+                while ((entry = jarInputStream.getNextJarEntry()) != null) {
+                    if (entry.isDirectory() || !entry.getName().endsWith(".class")) {
+                        continue;
+                    }
+
+                    String className = entry.getName().replaceAll("/", ".").replaceAll(".class", "");
+                    HashingInputStream inputStream = new HashingInputStream(Hashing.sha256(), jarInputStream);
+                    ByteStreams.exhaust(inputStream);
+
+                    multiMapPut(classToJars,
+                            className,
+                            resolvedArtifact.getModuleVersion().getId());
+
+                    multiMapPut(tempClassToHashCodes,
+                            className,
+                            inputStream.hash());
+                }
+            } catch (IOException e) {
+                log.error("Failed to read JarFile {}", resolvedArtifact, e);
+                throw new RuntimeException(e);
+            }
+        });
+
+        // discard all the classes that only come from one jar - these are completely safe!
+        classToJars.entrySet().stream()
```

Comment:
don't think we need this collection -- finding which entries of `tempClassToHashCodes` have value of size > 1 (which we're doing a few lines down) should be sufficient

Reply:
oh whoops I didn't read this closely enough